geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/QueryInterceptor.scala

```diff
@@ -0,0 +1,122 @@
+/***********************************************************************
+ * Copyright (c) 2013-2019 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index.planning
+
+import java.io.Closeable
+import java.util.concurrent.TimeUnit
+
+import com.github.benmanes.caffeine.cache.{CacheLoader, Caffeine}
+import com.typesafe.scalalogging.LazyLogging
+import org.geotools.data.{DataStore, Query}
+import org.locationtech.geomesa.index.metadata.CachedLazyMetadata
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes.Configs
+import org.locationtech.geomesa.utils.io.CloseWithLogging
+import org.opengis.feature.simple.SimpleFeatureType
+
+import scala.util.control.NonFatal
+
+/**
+  * Provides a hook to modify a query before executing it
+  */
+trait QueryInterceptor extends Closeable {
```

Comment:
Do we expected folks to implement this from Java or Scala?  (I didn't know if we had a strong preference for Java interop or not.)

Reply:
i think this class actually would be easily extensible in java. In general I would say we can use scala for most things... has anyone actually ever extended any of our java apis?